### PR TITLE
[Notifier] Add RingCentral bridge

### PR DESCRIPTION
**Pull request by gnito-org**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/17485
| Recipe         | https://github.com/symfony/recipes/pull/1155

This PR add the [RingCentral](https://www.ringcentral.com) Notifier bridge.

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/RingCentral/.gitattributes [hunk 1]
```diff
@@ -0,0 +1,4 @@
+/Tests export-ignore
+/phpunit.xml.dist export-ignore
+/.gitattributes export-ignore
+/.gitignore export-ignore
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/.gitattributes]: Please add a newline to remove the red pictogram, thanks

File: src/Symfony/Component/Notifier/Bridge/RingCentral/.gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+vendor/
+composer.lock
+phpunit.xml
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/.gitignore]: Please add a newline to remove the red pictogram, thanks

File: src/Symfony/Component/Notifier/Bridge/RingCentral/LICENSE [hunk 3]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2019-2022 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/LICENSE]:
```suggestion
Copyright (c) 2022 Fabien Potencier
```

File: src/Symfony/Component/Notifier/Bridge/RingCentral/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+-----------
+
+```
+RINGCENTRAL_DSN=ringcentral://API_TOKEN@default?from=FROM
+```
+
+where:
+
+- `API_TOKEN` is your Ring Central OAuth 2 token
+- `FROM` is your sender
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/README.md]:
```suggestion
 - `API_TOKEN` is your Ring Central OAuth 2 token
 - `FROM` is your sender
```

File: src/Symfony/Component/Notifier/Bridge/RingCentral/RingCentralOptions.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    public function setCountryCallingCode(string $countryCallingCode): self
+    {
+        $this->options['country_calling_code'] = $countryCallingCode;
+        return $this;
+    }
+
+    public function setCountryId(string $countryId): self
+    {
+        $this->options['country_id'] = $countryId;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/RingCentralOptions.php:72]:
```suggestion
        $this->options['country_id'] = $countryId;

```

File: src/Symfony/Component/Notifier/Bridge/RingCentral/RingCentralTransport.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        return sprintf('ringcentral://%s?from=%s', $this->getEndpoint(), $this->from);
+    }
+
+    public function supports(MessageInterface $message): bool
+    {
+        return $message instanceof SmsMessage;
+    }
+
+    protected function doSend(MessageInterface $message): SentMessage
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/RingCentralTransport.php:51]: Please adjust like described in #48389

File: src/Symfony/Component/Notifier/Bridge/RingCentral/RingCentralTransportFactory.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+use Symfony\Component\Notifier\Transport\Dsn;
+
+/**
+ * @author gnito-org <https://github.com/gnito-org>
+ */
+final class RingCentralTransportFactory extends AbstractTransportFactory
+{
+    private const TRANSPORT_SCHEME = 'ringcentral';
+
+    public function create(Dsn $dsn): RingCentralTransport
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/RingCentralTransportFactory.php:25]: Please adjust like described in #48389

File: src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportFactoryTest.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+final class RingCentralTransportFactoryTest extends TransportFactoryTestCase
+{
+    public function createFactory(): RingCentralTransportFactory
+    {
+        return new RingCentralTransportFactory();
+    }
+
+    public function createProvider(): iterable
+    {
+        yield [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportFactoryTest.php]: One line

File: src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportFactoryTest.php [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public function missingRequiredOptionProvider(): iterable
+    {
+        yield 'missing option: from' => ['ringcentral://apiToken@default'];
+    }
+
+    public function supportsProvider(): iterable
+    {
+        yield [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportFactoryTest.php]: One line

File: src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportFactoryTest.php [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+            true,
+            'ringcentral://apiToken@default?from=0611223344',
+        ];
+        yield [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportFactoryTest.php]: One line

File: src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+use Symfony\Component\Notifier\Message\ChatMessage;
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Test\TransportTestCase;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+final class RingCentralTransportTest extends TransportTestCase
+{
+    public function createTransport(
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php]: One line

File: src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        HttpClientInterface $client = null,
+        string $from = 'from'
+    ): RingCentralTransport {
+        return new RingCentralTransport('apiToken', $from, $client ?? $this->createMock(HttpClientInterface::class));
+    }
+
+    public function invalidFromProvider(): iterable
+    {
+        // phone numbers
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        yield 'no zero at start if phone number' => ['+0'];
+        yield 'phone number too short' => ['+1'];
+    }
+
+    public function supportedMessagesProvider(): iterable
+    {
+        yield [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php]: One line

File: src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+    public function unsupportedMessagesProvider(): iterable
+    {
+        yield [new ChatMessage('Hello!')];
+        yield [$this->createMock(MessageInterface::class)];
+    }
+
+    public function validFromProvider(): iterable
+    {
+        // phone numbers
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+            }
+        );
+        $transport = $this->createTransport($client, $from);
+        $sentMessage = $transport->send($message);
+        self::assertSame('foo', $sentMessage->getMessageId());
+    }
+
+    public function toStringProvider(): iterable
+    {
+        yield [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/Tests/RingCentralTransportTest.php]: One line

File: src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json [hunk 16]
```diff
@@ -0,0 +1,35 @@
+{
+    "name": "symfony/ring-central-notifier",
+    "type": "symfony-notifier-bridge",
+    "description": "Symfony RingCentral Notifier Bridge",
+    "keywords": [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json:5]: One line

File: src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=8.1",
+        "symfony/http-client": "^5.4|^6.0",
+        "symfony/notifier": "^5.4|^6.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json]:
```suggestion
        "symfony/notifier": "^6.3"
```

File: src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    },
+    "autoload": {
+        "psr-4": {
+            "Symfony\\Component\\Notifier\\Bridge\\RingCentral\\": ""
+        },
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json]: One line

File: src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            "name": "gnito-org",
+            "homepage": "https://github.com/gnito-org"
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json:21]: Please add event-dispatcher to require-dev

File: src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php [hunk 20]
```diff
@@ -140,6 +140,10 @@ class UnsupportedSchemeException extends LogicException
             'class' => Bridge\OvhCloud\OvhCloudTransportFactory::class,
             'package' => 'symfony/ovh-cloud-notifier',
         ],
+        'ringcentral' => [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php:147]: should be named
```suggestion
        'ring-central' => [
```

Am I right @fabpot?

File: src/Symfony/Component/Notifier/Bridge/RingCentral/README.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+ - `API_TOKEN` is your Ring Central OAuth 2 token
+ - `FROM` is your sender
+
+Resources
+---------
+
+* [Contributing](https://symfony.com/doc/current/contributing/index.html)
+* [Report issues](https://github.com/symfony/symfony/issues) and
+  [send Pull Requests](https://github.com/symfony/symfony/pulls)
+  in the [main Symfony repository](https://github.com/symfony/symfony)
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/README.md]:
```suggestion
 * [Contributing](https://symfony.com/doc/current/contributing/index.html)
 * [Report issues](https://github.com/symfony/symfony/issues) and
   [send Pull Requests](https://github.com/symfony/symfony/pulls)
   in the [main Symfony repository](https://github.com/symfony/symfony)
```

File: src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+        "php": ">=8.1",
+        "symfony/event-dispatcher": "^5.4|^6.0",
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/RingCentral/composer.json]: Please add event-dispatcher to require-dev

File: src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php [hunk 23]
```diff
@@ -140,6 +140,10 @@ class UnsupportedSchemeException extends LogicException
             'class' => Bridge\OvhCloud\OvhCloudTransportFactory::class,
             'package' => 'symfony/ovh-cloud-notifier',
         ],
+        'ring-central' => [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php]:
```suggestion
        'ringcentral' => [
```

sorry, this is the scheme and it should not contain a dash, my fault

**Review by OskarStark — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @gnito-org.

**Pull request merged**